{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert output of munge_sumstats.py to matlab format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools as it\n",
    "import pandas as pd\n",
    "import scipy.io as sio\n",
    "\n",
    "# complementary bases\n",
    "COMPLEMENT = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "# bases\n",
    "BASES = COMPLEMENT.keys()\n",
    "# true iff strand ambiguous\n",
    "STRAND_AMBIGUOUS = {''.join(x): x[0] == COMPLEMENT[x[1]]\n",
    "                    for x in it.product(BASES, BASES)\n",
    "                    if x[0] != x[1]}\n",
    "# SNPS we want to keep (pairs of alleles)\n",
    "VALID_SNPS = {x for x in map(lambda y: ''.join(y), it.product(BASES, BASES))\n",
    "              if x[0] != x[1] and not STRAND_AMBIGUOUS[x]}\n",
    "# T iff SNP 1 has the same alleles as SNP 2 (allowing for strand or ref allele flip).\n",
    "MATCH_ALLELES = {x for x in map(lambda y: ''.join(y), it.product(VALID_SNPS, VALID_SNPS))\n",
    "                 # strand and ref match\n",
    "                 if ((x[0] == x[2]) and (x[1] == x[3])) or\n",
    "                 # ref match, strand flip\n",
    "                 ((x[0] == COMPLEMENT[x[2]]) and (x[1] == COMPLEMENT[x[3]])) or\n",
    "                 # ref flip, strand match\n",
    "                 ((x[0] == x[3]) and (x[1] == x[2])) or\n",
    "                 ((x[0] == COMPLEMENT[x[3]]) and (x[1] == COMPLEMENT[x[2]]))}  # strand and ref flip\n",
    "# T iff SNP 1 has the same alleles as SNP 2 w/ ref allele flip.\n",
    "FLIP_ALLELES = {''.join(x):\n",
    "                ((x[0] == x[3]) and (x[1] == x[2])) or  # strand match\n",
    "                # strand flip\n",
    "                ((x[0] == COMPLEMENT[x[3]]) and (x[1] == COMPLEMENT[x[2]]))\n",
    "                for x in MATCH_ALLELES}\n",
    "\n",
    "def _filter_alleles(alleles):\n",
    "    '''Remove bad variants (mismatched alleles, non-SNPs, strand ambiguous).'''\n",
    "    ii = alleles.apply(lambda y: y in MATCH_ALLELES)\n",
    "    return ii\n",
    "\n",
    "class Logger(object):\n",
    "    '''\n",
    "    Lightweight logging.\n",
    "    TODO: replace with logging module\n",
    "    '''\n",
    "    def __init__(self, fh = None):\n",
    "        self.log_fh = open(fh, 'wb') if fh is not None else None\n",
    "\n",
    "    def log(self, msg):\n",
    "        '''\n",
    "        Print to log file and stdout with a single command.\n",
    "        '''\n",
    "        if self.log_fh is not None: print >>self.log_fh, msg\n",
    "        print(msg)\n",
    "\n",
    "def read_csv(fh, **kwargs):\n",
    "    return pd.read_csv(fh, delim_whitespace=True, na_values='.', **kwargs)\n",
    "\n",
    "def read_sumstats(fh, alleles=False, dropna=True):\n",
    "    '''Parses .sumstats files. See docs/file_formats_sumstats.txt.'''\n",
    "    dtype_dict = {'SNP': str,   'Z': float, 'N': float, 'A1': str, 'A2': str}\n",
    "    compression = 'infer'  # get_compression(fh)\n",
    "    usecols = ['SNP', 'Z', 'N']\n",
    "    if alleles:\n",
    "        usecols += ['A1', 'A2']\n",
    "\n",
    "    try:\n",
    "        x = read_csv(fh, usecols=usecols, dtype=dtype_dict, compression=compression)\n",
    "    except (AttributeError, ValueError) as e:\n",
    "        raise ValueError('Improperly formatted sumstats file: ' + str(e.args))\n",
    "\n",
    "    if dropna:\n",
    "        x = x.dropna(how='any')\n",
    "\n",
    "    return x\n",
    "\n",
    "def _read_sumstats(log, fh, alleles=False, dropna=False):\n",
    "    '''Parse summary statistics.'''\n",
    "    log.log('Reading summary statistics from {S} ...'.format(S=fh))\n",
    "    sumstats = read_sumstats(fh, alleles=alleles, dropna=dropna)\n",
    "    log_msg = 'Read summary statistics for {N} SNPs.'\n",
    "    log.log(log_msg.format(N=len(sumstats)))\n",
    "    m = len(sumstats)\n",
    "    sumstats = sumstats.drop_duplicates(subset='SNP')\n",
    "    if m > len(sumstats):\n",
    "        log.log(\n",
    "            'Dropped {M} SNPs with duplicated rs numbers.'.format(M=m - len(sumstats)))\n",
    "\n",
    "    return sumstats\n",
    "\n",
    "def _select_and_log(x, ii, log, msg):\n",
    "    '''Fiter down to rows that are True in ii. Log # of SNPs removed.'''\n",
    "    new_len = ii.sum()\n",
    "    if new_len == 0:\n",
    "        raise ValueError(msg.format(N=0))\n",
    "    else:\n",
    "        x = x[ii]\n",
    "        log.log(msg.format(N=new_len))\n",
    "    return x\n",
    "\n",
    "def _align_alleles(z, alleles):\n",
    "    '''Align Z1 and Z2 to same choice of ref allele (allowing for strand flip).'''\n",
    "    try:\n",
    "        z *= (-1) ** alleles.apply(lambda y: FLIP_ALLELES[y])\n",
    "    except KeyError as e:\n",
    "        msg = 'Incompatible alleles in .sumstats files: %s. ' % e.args\n",
    "        msg += 'Did you forget to use --merge-alleles with munge_sumstats.py?'\n",
    "        raise KeyError(msg)\n",
    "    return z\n",
    "\n",
    "def sumstats_to_mat(reffile, sumstats, out):\n",
    "    log = Logger();\n",
    "    \n",
    "    log.log('Reading reference SNPs from {S} ...'.format(S=reffile))\n",
    "    ref = pd.read_csv(reffile, delim_whitespace=True)\n",
    "    ref = ref[['SNP','A1','A2']]\n",
    "    log.log('Read reference data for {N} SNPs.'.format(N=len(ref)))\n",
    "    \n",
    "    # Reading summary statistics\n",
    "    s = _read_sumstats(log, sumstats, alleles=True, dropna=False)\n",
    "    s = s.rename(columns={'A1':'A1x','A2':'A2x'})\n",
    "    \n",
    "    # Merging summary statistics with reference file\n",
    "    s = pd.merge(ref, s, how='left', on=['SNP'])\n",
    "    log.log('After merging with {F}, {N} SNPs remain.'.format(F='reference file', N=len(s)))\n",
    "    \n",
    "    # Drop n/a and mismatched alleles\n",
    "    s = s.dropna(how='any')\n",
    "    log.log('After dropping n/a, {N} SNPs remain.'.format(F='ref', N=len(s)))\n",
    "    s = _select_and_log(s, _filter_alleles(s.A1 + s.A2 + s.A1x + s.A2x), Logger(), '{N} SNPs with valid alleles.')\n",
    "    \n",
    "    # Flip Z score to match alleles\n",
    "    s.Z = _align_alleles(s.Z, s.A1 + s.A2 + s.A1x + s.A2x)\n",
    "\n",
    "    # Align again to the reference (adds nan for missing SNPs)\n",
    "    s = pd.merge(ref[['SNP']], s[['SNP', 'Z', 'N']], how='left', on=['SNP'])\n",
    "\n",
    "    # Save result to .mat file\n",
    "    save_dict = {'zvec':s['Z'].values, 'nvec':s['N'].values}\n",
    "    sio.savemat(out, save_dict, format='5', do_compression=False, oned_as='column', appendmat=False)\n",
    "    print('{0} zvec and nvec values, {1} of which are not null, written to {f}'.format(len(s), (~s['Z'].isnull()).sum(), f=out))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading reference SNPs from 1m.ref.gz ...\n",
      "Read reference data for 1190321 SNPs.\n",
      "Reading summary statistics from BIP.sumstats.gz ...\n",
      "Read summary statistics for 1125109 SNPs.\n",
      "After merging with reference file, 1190321 SNPs remain.\n",
      "After dropping n/a, 1115633 SNPs remain.\n",
      "1115632 SNPs with valid alleles.\n",
      "1190321 zvec and nvec values, 1115632 of which are not null, written to BIP.mat\n",
      "Reading reference SNPs from 1m.ref.gz ...\n",
      "Read reference data for 1190321 SNPs.\n",
      "Reading summary statistics from SCZ.sumstats.gz ...\n",
      "Read summary statistics for 1135278 SNPs.\n",
      "After merging with reference file, 1190321 SNPs remain.\n",
      "After dropping n/a, 1125702 SNPs remain.\n",
      "1125701 SNPs with valid alleles.\n",
      "1190321 zvec and nvec values, 1125701 of which are not null, written to SCZ.mat\n",
      "Reading reference SNPs from 1m.ref.gz ...\n",
      "Read reference data for 1190321 SNPs.\n",
      "Reading summary statistics from PTSD.sumstats.gz ...\n",
      "Read summary statistics for 1078817 SNPs.\n",
      "After merging with reference file, 1190321 SNPs remain.\n",
      "After dropping n/a, 1077567 SNPs remain.\n",
      "1077567 SNPs with valid alleles.\n",
      "1190321 zvec and nvec values, 1077567 of which are not null, written to PTSD.mat\n",
      "Reading reference SNPs from 1m.ref.gz ...\n",
      "Read reference data for 1190321 SNPs.\n",
      "Reading summary statistics from MDD.sumstats.gz ...\n",
      "Read summary statistics for 1086802 SNPs.\n",
      "After merging with reference file, 1190321 SNPs remain.\n",
      "After dropping n/a, 1084345 SNPs remain.\n",
      "1084345 SNPs with valid alleles.\n",
      "1190321 zvec and nvec values, 1084345 of which are not null, written to MDD.mat\n"
     ]
    }
   ],
   "source": [
    "reffile = r'1m.ref.gz'\n",
    "sumstats_to_mat(reffile, r'BIP.sumstats.gz', 'BIP.mat')\n",
    "sumstats_to_mat(reffile, r'SCZ.sumstats.gz', 'SCZ.mat')\n",
    "sumstats_to_mat(reffile, r'PTSD.sumstats.gz', 'PTSD.mat')\n",
    "sumstats_to_mat(reffile, r'MDD.sumstats.gz', 'MDD.mat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
